### PR TITLE
Disable fusion for impure externs

### LINKED ISSUE #9943: [js] analyzer-optimize can re-order api (extern) calls what can cause a mess
Haxe ver: 4.1.4, 3.4.4
Sample code:
https://try.haxe.org/index.php#8888E
```
class Test {
    static function main() {
                var document = js.Browser.window.document;
		var e_div_ = document.createElement("div");
		e_div_.style.position = "absolute";
		e_div_.style.left = "0";
		e_div_.style.top = "0";
		e_div_.style.width = "512px";
		e_div_.style.height = "256px";
		e_div_.style.visibility = "hidden";
		e_div_.style.overflow = 'hidden';
		var e_span_ = document.createElement("span");
		e_span_.innerText = "|Yy";
		var e_block_ = document.createElement("div");
		e_block_.style.width = "1px";
		e_block_.style.height = "0";
		e_block_.style.display = "inline-block";
		e_div_.appendChild(e_span_);
		e_div_.appendChild(e_block_);
                document.body.appendChild(e_div_);
		e_block_.style.verticalAlign = "baseline";
		var a = (e_block_.offsetTop - e_span_.offsetTop);
		e_block_.style.verticalAlign = "bottom";
		var h = (e_block_.offsetTop - e_span_.offsetTop);
		e_block_.style.verticalAlign = "middle";
		var m = (e_block_.offsetTop - e_span_.offsetTop);
                trace('a=$a, h=$h, m=$m');
    }
}
```
Bad result with analyzer-optimize:
``a=10, h=10, m=10``
Part of js code:
```
	e_block_.style.verticalAlign = "baseline";
	e_block_.style.verticalAlign = "bottom";
	e_block_.style.verticalAlign = "middle";
	console.log("a=" + (e_block_.offsetTop - e_span_.offsetTop) + ", h=" + (e_block_.offsetTop - e_span_.offsetTop) + ", m=" + (e_block_.offsetTop - e_span_.offsetTop));
```
Good result w/o analyzer-optimize:
``a=14, h=18, m=10``
Part of js code:
```
	e_block_.style.verticalAlign = "baseline";
	var a = e_block_.offsetTop - e_span_.offsetTop;
	e_block_.style.verticalAlign = "bottom";
	var h = e_block_.offsetTop - e_span_.offsetTop;
	e_block_.style.verticalAlign = "middle";
	var m = e_block_.offsetTop - e_span_.offsetTop;
	console.log("a=" + a + ", h=" + h + ", m=" + m);
```

### PR DESCRIPTION
**Pull request by RblSb**
Pretty scary bug, i think.

There is 5 test failures:
```
unit.TestExceptions
  testExceptionStack: FAILURE .FFFFF.....
    line: 255, _without_ throws: {
	file : /home/runner/work/haxe/haxe/tests/unit/bin/js/-Dgithub-DLinux-Djs-classic/unit.js, 
	line : NaN, 
	method : stacksWithoutThrowLevel2
} is expected, but got {

}
    line: 255, _without_ throws: {
	file : /home/runner/work/haxe/haxe/tests/unit/bin/js/-Dgithub-DLinux-Djs-classic/unit.js, 
	line : NaN, 
	method : stacksWithoutThrowLevel2
} is expected, but got {

}
    line: 255, _without_ throws: {
	file : /home/runner/work/haxe/haxe/tests/unit/bin/js/-Dgithub-DLinux-Djs-classic/unit.js, 
	line : NaN, 
	method : stacksWithoutThrowLevel2
} is expected, but got {

}
    line: 255, _without_ throws: {
	file : /home/runner/work/haxe/haxe/tests/unit/bin/js/-Dgithub-DLinux-Djs-classic/unit.js, 
	line : NaN, 
	method : stacksWithoutThrowLevel2
} is expected, but got {

}
    line: 255, _without_ throws: {
	file : /home/runner/work/haxe/haxe/tests/unit/bin/js/-Dgithub-DLinux-Djs-classic/unit.js, 
	line : NaN, 
	method : stacksWithoutThrowLevel2
} is expected, but got {

}
```

Please review [this test](https://github.com/HaxeFoundation/haxe/blob/development/tests/unit/src/unit/TestExceptions.hx#L235) and tryHaxeStack js impl, because diff change seems like a fix for Haxe code:
[Haxe tryHaxeStack function](https://github.com/HaxeFoundation/haxe/blob/development/std/js/_std/haxe/NativeStackTrace.hx#L89)
And unit.js diff for her:
```diff
haxe_NativeStackTrace.tryHaxeStack = function(e) {
	if(e == null) {
		return [];
	}
	var oldValue = Error.prepareStackTrace;
	Error.prepareStackTrace = haxe_NativeStackTrace.prepareHxStackTrace;
+	var stack = e.stack;
	Error.prepareStackTrace = oldValue;
-	return e.stack;
+	return stack;
};
```

closes #9943

### CODE REVIEW

**Comment by Simn** [comment]
Change makes sense to me. @RealyUniqueName Could you check the failing tests?

**Comment by RealyUniqueName** [comment]
@RblSb 30d639bca164e6d00dbd3a1d342f34b149c5945c fixes the issue. Merge it to your branch.

**Comment by RblSb** [comment]
Diff for my SyncTube project (~9000 generated LoC). Looks fair for me. Can be improved with `@:pure` on extern classes with extern generators patches. Current html extern generator handles pure js stuff from Web IDLs btw, these idls just poor about pure stuff for now.
```diff
nodejs server:

-                       var events = JSON.parse(js_node_Fs.readFileSync(path,{ encoding : "utf8"}));
+                       var text = js_node_Fs.readFileSync(path,{ encoding : "utf8"});
+                       var events = JSON.parse(text);

-               res.end("File " + js_node_Path.relative(server_HttpServer.dir,filePath) + " not found.");
+               var rel = js_node_Path.relative(server_HttpServer.dir,filePath);
+               res.end("File " + rel + " not found.");

-       js_node_Fs.createReadStream(filePath,{ start : start, end : end}).pipe(res);
+       var videoStream = js_node_Fs.createReadStream(filePath,{ start : start, end : end});
+       videoStream.pipe(res);

-       var proxy = server_HttpServer.proxyRequest(StringTools.replace(req.url,"/proxy?url=",""),req,res,function(proxyReq) {
+       var url = StringTools.replace(req.url,"/proxy?url=","");
+       var proxy = server_HttpServer.proxyRequest(url,req,res,function(proxyReq) {

-       var proxy = (url1.protocol == "https:" ? js_node_Https.request : js_node_Http.request)({ host : url1.hostname, port : Std.parseInt(url1.port), path : url1.pathname + url1.search, method : req.method},function(proxyReq) {
+       var options = { host : url1.hostname, port : Std.parseInt(url1.port), path : url1.pathname + url1.search, method : req.method};
+       var request = url1.protocol == "https:" ? js_node_Https.request : js_node_Http.request;
+       var proxy = request(options,function(proxyReq) {

+               var isAdmin = this.config.localAdmins && req.socket.localAddress == ip;
                var client = new Client(ws,req,id,name,0);
-               client.setGroupFlag(ClientGroup.Admin,this.config.localAdmins && req.socket.localAddress == ip);
+               client.setGroupFlag(ClientGroup.Admin,isAdmin);

js client:

-               if(smilesBtn.classList.toggle("active")) {
+               var isActive = smilesBtn.classList.toggle("active");
+               if(isActive) {

-               if(client_Buttons.isElementEditable(e.target)) {
+               var target = e.target;
+               if(client_Buttons.isElementEditable(target)) {

        onKeyDown: function(e) {
-               switch(e.keyCode) {
+               var key = e.keyCode;
+               switch(key) {

-                       url = "" + protocol + "//" + $global.location.hostname + ":" + $global.location.port + url;
+                       var host = $global.location.hostname;
+                       var port = $global.location.port;
+                       url = "" + protocol + "//" + host + ":" + port + url;

-               this.setItemElementType(this.videoItemsEl.children[pos],this.videoList.items[this.videoList.pos].isTemp);
+               var el = this.videoItemsEl.children[pos];
+               this.setItemElementType(el,this.videoList.items[this.videoList.pos].isTemp);

-       return client_Settings.checkData(JSON.parse(client_Settings.storage.getItem("data")));
+       var data = JSON.parse(client_Settings.storage.getItem("data"));
+       return client_Settings.checkData(data);

-       trackEl.track.mode = "showing";
+       var track = trackEl.track;
+       track.mode = "showing";
```

File: src/optimization/analyzerTexpr.ml [hunk 1]
```diff
@@ -590,14 +590,32 @@ module Fusion = struct
 			let can_be_used_as_value = can_be_used_as_value com e in
 			let is_compiler_generated = match v.v_kind with VUser _ | VInlined -> false | _ -> true in
 			let has_type_params = match v.v_extra with Some ve when ve.v_params <> [] -> true | _ -> false in
+			let rec contains_pure_meta metadata = match metadata with
```
Comment (Simn) [inline src/optimization/analyzerTexpr.ml]: Check out the functions in purityState.ml, you should be able to use those.

**Comment by RblSb** [comment]
Also without Meta.NoClosure check:
```diff
-               js_node_Fs.writeFileSync("" + crashesFolder + "/" + (DateTools.format(new Date(),"%Y-%m-%d_%H_%M_%S") + "-" + type) + ".json",JSON.stringify(data,null,"\t"));
+               var name = DateTools.format(new Date(),"%Y-%m-%d_%H_%M_%S") + "-" + type;
+               js_node_Fs.writeFileSync("" + crashesFolder + "/" + name + ".json",JSON.stringify(data,null,"\t"));

-               window.visualViewport.addEventListener("resize",function(e) {
+               var viewport = window.visualViewport;
+               viewport.addEventListener("resize",function(e) {
```
Without `Meta.CoreApi` check there is little more purity fails, don't know if it should be removed too.

**Comment by Simn** [comment]
Thank you!

**Pull request merged**